Source/WebCore/css/StyleRule.h

```diff
@@ -149,32 +135,40 @@ class StyleRule final : public StyleRuleBase {
 
     static unsigned averageSizeInBytes();
     void setProperties(Ref<StyleProperties>&&);
-    void setNestedRules(Vector<Ref<StyleRuleBase>>);
-    void setResolvedSelectorList(CSSSelectorList&&) const;
-    const Vector<Ref<StyleRuleBase>>& nestedRules() const;
-    void appendNestedRule(Ref<StyleRuleBase>&& rule) { rareData().nestedRules.append(WTFMove(rule)); }
 
-private:
-    StyleRule(Ref<StyleProperties>&&, bool hasDocumentSecurityOrigin, CSSSelectorList&&, Vector<Ref<StyleRuleBase>>&&);
-    StyleRule(bool hasDocumentSecurityOrigin, CSSSelectorList&&);
+protected:
+    StyleRule(Ref<StyleProperties>&&, bool hasDocumentSecurityOrigin, CSSSelectorList&&);
     StyleRule(const StyleRule&);
 
+private:
     static Ref<StyleRule> createForSplitting(const Vector<const CSSSelector*>&, Ref<StyleProperties>&&, bool hasDocumentSecurityOrigin);
 
-    StyleRuleRareData& rareData() const;
-
     bool m_isSplitRule { false };
     bool m_isLastRuleInSplitRule { false };
 
     mutable Ref<StyleProperties> m_properties;
     CSSSelectorList m_selectorList;
-    mutable std::unique_ptr<StyleRuleRareData> m_rareData;
-
 #if ENABLE(CSS_SELECTOR_JIT)
     mutable UniqueArray<CompiledSelector> m_compiledSelectors;
 #endif
 };
 
+class StyleWithNestingRule final : public StyleRule {
```

Comment:
What is the plan for the (rare but possible I think) case where you have a regular StyleRule from the parser and then someone uses CSSOM to add a nested child rule to it?

Reply:
It's not possible for the moment (we haven't plug the `insertRule` and we don't parse `&` in top-level selector).
But we will have to deal with it when implementing complete CSSOM support indeed.